File: app/src/main/java/ir/ghararemaghzha/game/firebase/MyFirebaseMessagingService.java
package ir.ghararemaghzha.game.firebase;

import android.content.Intent;

import androidx.annotation.NonNull;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import java.util.Map;

import io.realm.Realm;
import ir.ghararemaghzha.game.classes.MySettingsPreference;
import ir.ghararemaghzha.game.classes.MySharedPreference;
import ir.ghararemaghzha.game.classes.Utils;
import ir.ghararemaghzha.game.models.MessageModel;

import static ir.ghararemaghzha.game.classes.Const.GHARAREHMAGHZHA_BROADCAST;
import static ir.ghararemaghzha.game.classes.Const.GHARAREHMAGHZHA_BROADCAST_SUPPORT_EXTRA;
import static ir.ghararemaghzha.game.classes.Utils.getNextKey;

public class MyFirebaseMessagingService extends FirebaseMessagingService {

    @Override
    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
        Map<String, String> data = remoteMessage.getData();
        String title = data.get("title") + "";
        String notificationBody = data.get("body") + "";
        String clickAction = data.get("click_action") + "";
        Intent intent = new Intent();
        intent.setAction(GHARAREHMAGHZHA_BROADCAST);
        intent.putExtra(GHARAREHMAGHZHA_BROADCAST_SUPPORT_EXTRA, "new");

        if (!title.equals("support")) {
            String body = data.get("body") + "";
            String sender = data.get("sender") + "";
            MessageModel model = new MessageModel();
            String date = data.get("time");
            model.setStat(1);
            model.setMessage(body);
            model.setTitle(title);
            model.setSender(sender);
            model.setDate(date);
            model.setRead(0);
            Realm db = Realm.getDefaultInstance();
            model.setMessageId(getNextKey(db));
            db.beginTransaction();
            db.insert(model);
            db.commitTransaction();
        } else {
            MySharedPreference.getInstance(this).setUnreadChats(MySharedPreference.getInstance(this).getUnreadChats() + 1);
        }
        sendBroadcast(intent);
        if (MySettingsPreference.getInstance(this).getNotification())
            Utils.createNotification(this, title, notificationBody, clickAction);

    }


    @Override
    public void onNewToken(@NonNull String s) {
        super.onNewToken(s);
        if (!s.isEmpty())
            MySharedPreference.getInstance(this).setFbToken(s);
    }

}

File: app/src/main/java/ir/ghararemaghzha/game/firebase/MyFirebaseMessagingService.kt
package ir.ghararemaghzha.game.firebase

import android.content.Intent
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import io.realm.Realm
import ir.ghararemaghzha.game.classes.Const.GHARAREHMAGHZHA_BROADCAST
import ir.ghararemaghzha.game.classes.Const.GHARAREHMAGHZHA_BROADCAST_END
import ir.ghararemaghzha.game.classes.Const.GHARAREHMAGHZHA_BROADCAST_SUPPORT_EXTRA
import ir.ghararemaghzha.game.classes.MySettingsPreference
import ir.ghararemaghzha.game.classes.MySharedPreference
import ir.ghararemaghzha.game.classes.Utils
import ir.ghararemaghzha.game.models.MessageModel

class MyFirebaseMessagingService : FirebaseMessagingService() {

    override fun onNewToken(p0: String) {
        super.onNewToken(p0)
        if (p0.isNotEmpty())
            MySharedPreference.getInstance(this).setFbToken(p0)
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)
        val data = remoteMessage.data
        val title = data["title"] ?: ""
        var notificationBody = data["body"] ?: ""
        val clickAction = data["click_action"] ?: ""
        val intent = Intent()
        intent.action = GHARAREHMAGHZHA_BROADCAST
        intent.putExtra(GHARAREHMAGHZHA_BROADCAST_SUPPORT_EXTRA, "new")

        if (title.isNotEmpty() && notificationBody.isNotEmpty()) {

            if (title == "ended" && data["body"] == "ended") {
                val i = Intent()
                i.action = GHARAREHMAGHZHA_BROADCAST_END
                sendBroadcast(i)
                return
            }

            if (title != "support") {
                val body = data["body"]
                val sender = data["sender"]
                val model = MessageModel()
                val date = data["time"]
                model.stat = 1
                model.message = body ?: ""
                model.title = title
                model.sender = sender ?: ""
                model.date = date ?: ""
                model.read = 0
                val db = Realm.getDefaultInstance()
                model.messageId = Utils.getNextKey(db)
                db.beginTransaction()
                db.insert(model)
                db.commitTransaction()

            } else {
                MySharedPreference.getInstance(this).setUnreadChats(MySharedPreference.getInstance(this).getUnreadChats() + 1)
            }
            sendBroadcast(intent)
            if (MySettingsPreference.getInstance(this).getNotification()) {
                if (title != "support")
                    if (notificationBody.contains("<link>", true))
                        notificationBody = notificationBody.substring(0, notificationBody.indexOf("<link>") - 1)


                Utils.createNotification(this, title, notificationBody, clickAction)
            }
        }
    }
}
